crypto/engine/eng_devcrypto.c

```diff
@@ -202,24 +207,27 @@ static int cipher_do_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out,
     struct crypt_op cryp;
 #if !defined(COP_FLAG_WRITE_IV)
     unsigned char saved_iv[EVP_MAX_IV_LENGTH];
+    const unsigned char *ivptr;
+    unsigned char *iv = EVP_CIPHER_CTX_iv_noconst(ctx);
+    size_t nblocks, ivlen;
 #endif
 
     memset(&cryp, 0, sizeof(cryp));
     cryp.ses = cipher_ctx->sess.ses;
     cryp.len = inl;
     cryp.src = (void *)in;
     cryp.dst = (void *)out;
-    cryp.iv = (void *)EVP_CIPHER_CTX_iv_noconst(ctx);
+    cryp.iv = (void *)iv;
     cryp.op = cipher_ctx->op;
 #if !defined(COP_FLAG_WRITE_IV)
     cryp.flags = 0;
 
-    if (EVP_CIPHER_CTX_iv_length(ctx) > 0) {
-        assert(inl >= EVP_CIPHER_CTX_iv_length(ctx));
+    ivlen = EVP_CIPHER_CTX_iv_length(ctx);
+    if (ivlen > 0 && cipher_ctx->mode == EVP_CIPH_CBC_MODE) {
```

Comment:
Wouldn't it be better to do `... && cipher_ctx->mode != EVP_CIPH_CTR_MODE`?  I know that we *currently* only have support for CTR and CBC modes.

Come to think of it, you might want to do make a switch, to make future development easier...  something like this perhaps?

``` C
    if (ivlen > 0)
        switch (cipher_ctx->mode) {
        case EVP_CIPH_CTR_MODE:
            /* Do nothing here */
            break;
        default:
            assert(inl >= ivlen);
            if (!EVP_CIPHER_CTX_encrypting(ctx)) {
                unsigned char *ivptr = in + inl - EVP_CIPHER_CTX_iv_length(ctx);
                memcpy(saved_iv, ivptr, EVP_CIPHER_CTX_iv_length(ctx));
                ivptr = in + inl - ivlen;
                memcpy(saved_iv, ivptr, ivlen);
            }
        }
```

Reply:
We do support ECB.

Reply:
Yeah, but ECB usually doesn't have an IV (so IV length should be zero)

...

... and I just noticed that we specify a non-zero IV length in `cipher_data` for ECB cipher.  That's just plain wrong.

Reply:
I wondered about that, but didn't want to touch it before.

Reply:
Are you thinking you'll make that part of this PR, or would you rather that get corrected separately?  (I can do that)

Reply:
I do have a switch statement here.  Did I not push it?  I'll try again.